src/NuGet.Core/NuGet.Protocol/HttpSource/HttpCacheUtility.cs

```diff
@@ -67,21 +67,43 @@ public static class HttpCacheUtility
             // The update of a cached file is divided into two steps:
             // 1) Delete the old file.
             // 2) Create a new file with the same name.
+
+            // Some FileStream operations on Windows are synchronous even though it may not seem so.
+            // The HTTP stack rewrite in .NET Core 2.1 introduced circumstances whereby these
+            // synchronous FileStream calls will keep an IO completion thread busy and block other
+            // HTTP requests from completing.  The immediate solution is to perform write and read
+            // operations on separate streams, but only on .NET Core where the problem exists.
+            // See https://github.com/dotnet/corefx/issues/31914 for details.
+            const int writeBufferSize =
+#if IS_CORECLR
+                1;  // This disables write buffering.
+#else
+                BufferSize;
+#endif
+
             using (var fileStream = new FileStream(
                 result.NewFile,
                 FileMode.Create,
-                FileAccess.ReadWrite,
+                FileAccess.Write,
                 FileShare.None,
-                BufferSize,
+                writeBufferSize,
                 useAsync: true))
             {
                 using (var networkStream = await response.Content.ReadAsStreamAsync())
                 {
                     await networkStream.CopyToAsync(fileStream, BufferSize, cancellationToken);
                 }
+            }
 
+            using (var fileStream = new FileStream(
+                result.NewFile,
+                FileMode.Open,
+                FileAccess.Read,
+                FileShare.None,
+                BufferSize,
+                useAsync: true))
+            {
```

Comment:
Will it add any performance impact for desktop frameworks since it will always open separate streams to read and write? 

Or shouldn't we just restrict these separate streams for netcore?

Reply:
Dsicussed offline and got this resolved. Updated the PR description accordingly